sshuttle/ssnet.py

```diff
@@ -620,10 +647,61 @@ def got_packet(self, cmd, data):
         elif cmd == CMD_TCP_DATA:
             self.isWrite = True
             self.buf.append(data)
+            self.maybe_pause(data)
+        elif cmd == CMD_PAUSE:
+            self.isPaused = True
+            debug2('MuxWrapper.got_packet received CMD_PAUSE on channel %s\n' % self.channel)
+        elif cmd == CMD_RESUME:
+            self.isPaused = False
+            debug2('MuxWrapper.got_packet received CMD_RESUME on channel %s\n' % self.channel)
         else:
             raise Exception('unknown command %d (%d bytes)'
                             % (cmd, len(data)))
 
+    # Check size of overall buffer; if too big, and individual buffer too big, send a pause.
+    def maybe_pause(self, data):
+        self.buf_total = self.buf_total + len(data)
+
+        global _global_mux_wrapper_buffer_size
+        _global_mux_wrapper_buffer_size += len(data)
+
+        debug3('Global mux wrap buf size: %d. Individual buf size: %d. On channel %s\n'
+            % (_global_mux_wrapper_buffer_size, self.buf_total, self.channel))
+
+        if not self.isPaused and _global_mux_wrapper_buffer_size > PAUSE_TRAFFIC_TO_EDGE_THRESHOLD and self.buf_total\
+                > INDIVIDUAL_BUFFER_PAUSE_SIZE:
+            self.mux.send(self.channel, CMD_PAUSE, b(''))
```

Comment:
is the length of the pause command payload important to the protocol? If it is, then it might be worth introducing a self.mux.pause() that knows the payload.